flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java

```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
+            continue;
+         }
+         List<String> metricsNameList = assetIdAndMetricsNameList.get(assetID);
+         filterValueUnitsByMetricNames(valueUnits, metricsNameList);
+         assetAndValueUnitsMap.put(assetID, valueUnits);
+      }
+      return assetAndValueUnitsMap;
+   }
+
+   /**
+    * @param server
+    * @return
+    * Map<String, Map<String, String>>
+    * {
+    *    toName:{metricName,formula}
+    * }
+    *
+    * Sample value : SERVER_CONNECTED_PDU_CURRENT
+    * {
+    *    PDU:pduID|Current:{Current,id1+id2}
+    * }
+    */
+   private Map<String, Map<String, String>> getToNameAndFromulasMapForServer(Asset server){
+      Map<String, Map<String, String>> toNameAndFormulasMap = new HashMap<String, Map<String, String>>();
+      Map<String, String> formulas = server.getMetricsformulars();
+      if(formulas == null || formulas.isEmpty()) {
+         return toNameAndFormulasMap;
+      }
+      //Server Host Usage Formula
+      Map<String, String> hostUsageformulas = server.getMetricsformulars();
+      String hostUsageformulasInfo = formulas.get(FlowgateConstant.HOST_METRICS);
+      if (!StringUtils.isBlank(hostUsageformulasInfo)) {
+         hostUsageformulas =
+               server.metricsFormulaToMap(hostUsageformulasInfo, new TypeReference<Map<String, String>>(){});
+         for(Map.Entry<String, String> metricNameAndFormula : hostUsageformulas.entrySet()) {
+            Map<String, String> metricNameAndFormulaMap = new HashMap<>();
+            String toName = metricNameAndFormula.getKey();
+            metricNameAndFormulaMap.put(metricNameAndFormula.getKey(), metricNameAndFormula.getValue());
+            toNameAndFormulasMap.put(toName, metricNameAndFormulaMap);
+         }
+      }
+      //Server PDU formula
+      Map<String, Map<String, String>> pduFormulas = new HashMap<String, Map<String, String>>();
+      String pduFormulasInfo = formulas.get(FlowgateConstant.PDU);
+      pduFormulas =
+            server.metricsFormulaToMap(pduFormulasInfo, new TypeReference<Map<String, Map<String, String>>>() {});
+      Map<String,String> justficationfileds = server.getJustificationfields();
+      String allPduPortInfo = justficationfileds.get(FlowgateConstant.PDU_PORT_FOR_SERVER);
+      List<String> pduPorts = null;
+      Map<String, String> pduAssetIdAndUsedOutletMap = null;
+      if (!StringUtils.isEmpty(allPduPortInfo)) {
+         pduPorts = Arrays.asList(allPduPortInfo.split(FlowgateConstant.SPILIT_FLAG));
+         pduAssetIdAndUsedOutletMap = new HashMap<String, String>();
+         for (String pduPortInfo : pduPorts) {
+            // startport_FIELDSPLIT_endDeviceName_FIELDSPLIT_endport_FIELDSPLIT_endDeviceAssetID
+            // item[0] start port
+            // item[1] device name
+            // item[2] end port
+            // itme[3] assetid
+            String items[] = pduPortInfo.split(FlowgateConstant.SEPARATOR);
+            pduAssetIdAndUsedOutletMap.put(items[3], items[2]);
```

Comment:
consider server connect multiple outlet in same pdu.

Reply:
Done